### PR TITLE
feat: Approve multisig transaction

### PR DESCRIPTION
**Pull request by yanguoyu**
https://user-images.githubusercontent.com/12881040/161879267-1420fb4c-fcc9-4a50-ad30-753dfed1df23.mov

### CODE REVIEW

**Comment by Keith-CY** [comment]
https://user-images.githubusercontent.com/7271329/161929156-fcb77a26-989f-4008-9e16-0e3cfca91eb8.mov

I triggered an error on importing a transaction of multisig, it's a 0-of-2 multisig tx and 2 addresses are both of this wallet

**Comment by yanguoyu** [comment]
> Screen.Recording.2022-04-06.at.16.18.30.mov 
> I triggered an error on importing a transaction of multisig, it's a 0-of-2 multisig tx and 2 addresses are both of this wallet

It seems i should judge the json file has been tampered.

**Comment by Keith-CY** [comment]
conflicted

**Comment by yanguoyu** [comment]
> conflicted

Resolved

**Comment by yanguoyu** [comment]
> Screen.Recording.2022-04-06.at.16.18.30.mov 
> I triggered an error on importing a transaction of multisig, it's a 0-of-2 multisig tx and 2 addresses are both of this wallet

Fixed by commit https://github.com/nervosnetwork/neuron/pull/2324/commits/1b9ee21c9d937751f2b805dbf5e915c81e31b91a

File: packages/neuron-wallet/src/controllers/multisig.ts [hunk 1]
```diff
@@ -228,4 +229,25 @@ export default class MultisigController {
       result: balances
     }
   }
+
+  async loadMultisigTxJson(fullPayload: string) {
```
Comment (Keith-CY) [inline packages/neuron-wallet/src/controllers/multisig.ts:233]: Need more tests

File: packages/neuron-wallet/src/controllers/offline-sign.ts [hunk 2]
```diff
@@ -103,7 +102,11 @@ export default class OfflineSignController {
         transaction: tx
       })
 
-      signer.setStatus(SignStatus.Signed)
+      let signStatus = SignStatus.Signed
+      if (params.multisigConfig) {
+        signStatus = getMultisigStatus(params.multisigConfig, params.transaction.signatures)
```
Comment (Keith-CY) [inline packages/neuron-wallet/src/controllers/offline-sign.ts:107]: Should update test

File: packages/neuron-wallet/src/controllers/offline-sign.ts [hunk 3]
```diff
@@ -130,6 +133,20 @@ export default class OfflineSignController {
     return await this.exportTransactionAsJSON(signer.toJSON())
   }
 
+  public async signAndBroadcastTransaction(
```
Comment (Keith-CY) [inline packages/neuron-wallet/src/controllers/offline-sign.ts:136]: Should add tests

File: packages/neuron-ui/src/services/remote/offline.ts [hunk 4]
```diff
@@ -24,34 +25,22 @@ interface MultisigConfigs {
...
-interface Signatures {
+export interface Signatures {
   [hash: string]: string[]
 }
 
 export interface OfflineSignJSON {
-  transaction: any
+  transaction: {
+    signatures?: Signatures
+  } & Omit<any, 'signatures'>
```
[the remark this answers is not in the document]
Comment (yanguoyu) [reply packages/neuron-ui/src/services/remote/offline.ts:35]: I hope to pop up the type when i use transaction

[the remark this answers is not in the document]
Comment (Keith-CY) [reply packages/neuron-ui/src/services/remote/offline.ts:35]: > I hope to pop up `signatures ` when i use transaction

Is `{ signatures?: Signatures }` enough to indicate this type declaration?

**Comment by yanguoyu** [comment]
> Now that we're using optional chaining
> 
> ```
> if (onSubmitSuccess) {
>   onSubmitSuccess()
> }
> ```
> 
> could be more concise
> 
> ```
> onSubmitSuccess?.()
> ```

I delete `onSubmitSuccess` prop.

File: packages/neuron-ui/src/utils/getScriptFormat.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  {
+    CodeHash: string
+    HashType: string
+  }[]
+> = {
+  'SECP256K1/blake160': [DefaultLockInfo],
+  'SECP256K1/multisig': [MultiSigLockInfo],
+  anyone_can_pay: [AnyoneCanPayLockInfoOnLina, AnyoneCanPayLockInfoOnAggron],
+}
+
```
[the remark this answers is not in the document]
Comment (yanguoyu) [reply packages/neuron-ui/src/utils/getScriptFormat.ts:14]: I forgot delete it.

**Review by Keith-CY — APPROVED** [review]

File: packages/neuron-ui/src/services/remote/offline.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yanguoyu) [reply packages/neuron-ui/src/services/remote/offline.ts:35]: `transaction` include many property.I add `signatures ` property at this pr.I want to make sure `transaction` include `signatures ` property.

[the remark this answers is not in the document]
Comment (Keith-CY) [reply packages/neuron-ui/src/services/remote/offline.ts:35]: > `transaction` include many property.I add `signatures ` property at this pr.I want to make sure `transaction` include `signatures ` property.

Does
```
interface Transaction {
    signatures?: Signatures
    [key: string]: any
}
```
work?

**Comment by Keith-CY** [comment]
Conflicted

**Comment by yanguoyu** [comment]
> Conflicted

Resolved

**Comment by Keith-CY** [comment]
image


1. i18n is not handled correctly
2. imported tx has 1 signature and only 1 is missing, but the status indicates that 2 signature s are required.

**Comment by Keith-CY** [comment]
image
Throttle or loading should be added since `sign and export` takes time to allow the user clicking multiple times.

**Comment by yanguoyu** [comment]
https://github.com/nervosnetwork/neuron/pull/2324#issuecomment-1101543839 and https://github.com/nervosnetwork/neuron/pull/2324#issuecomment-1101547976 fixed by https://github.com/nervosnetwork/neuron/pull/2324/commits/c7118af1e267c52c060e3e9f09579da3fba4d5d4

File: packages/neuron-ui/src/components/ApproveMultisigTx/hooks.ts [hunk 6]
```diff
@@ -63,56 +64,82 @@ export const useSignAndExport = ({
...
+    const unspecifiedSignedAddresses: string[] = []
+    const unspecifiedUnsignedAddresses: string[] = []
+    for (let i = 0; i < multisigBlake160s.length; i++) {
+      const hasSigned = signatures?.[multisigLockHash]?.includes(multisigBlake160s[i])
+      if (i < multisigConfig.r) {
+        if (!hasSigned) {
+          specifiedUnsignedAddresses.push(multisigBlake160s[i])
+        }
+      } else {
+        ;(hasSigned ? unspecifiedSignedAddresses : unspecifiedUnsignedAddresses).push(multisigBlake160s[i])
```
Comment (Keith-CY) [inline packages/neuron-ui/src/components/ApproveMultisigTx/hooks.ts:104]: What does ` ? unspecifiedSignedAddress :` do?

File: packages/neuron-ui/src/components/ApproveMultisigTx/hooks.ts [hunk 7]
```diff
@@ -63,56 +64,82 @@ export const useSignAndExport = ({
...
     multisigConfig.addresses,
   ])
-  const [lackOfRCount, lackOfMCount] = useMemo(() => {
-    let [r, signed] = [0, 0]
-    for (let i = 0; i < multisigConfig.r; i++) {
-      if (!signatures?.[multisigLockHash]?.includes(multisigBlake160s[i])) {
-        r = multisigConfig.r - i
+  const addressBlake160s = useMemo(() => addresses.map(v => addressToScript(v.address).args), [addresses])
+  return useMemo(() => {
+    const notSpecifiedCount = multisigConfig.m - multisigConfig.r
```
Comment (Keith-CY) [inline packages/neuron-ui/src/components/ApproveMultisigTx/hooks.ts:93]: This code block is complex and better to add tests to indicate its rules.

File: packages/neuron-ui/src/components/ApproveMultisigTx/hooks.ts (hunk 6 above)
Comment (yanguoyu) [reply packages/neuron-ui/src/components/ApproveMultisigTx/hooks.ts:104]: This is convenient for calculate lack of unspecified signer. I wrote an extra semicolon, i will delete it.

**Review by Keith-CY — APPROVED** [review]

**Pull request merged**